dmoj/judge.py

```diff
@@ -289,12 +222,224 @@ def murder(self):
+            # won't get called if we exit the process right now (so we'd leak all files created by the grader). This
+            # should be refactored to have an explicit `cleanup()` or similar, rather than relying on refcounting
+            # working out.
+            self.grader = None
+
+    def _grade_cases(self):
+        problem = Problem(
+            self.submission.problem_id, self.submission.time_limit, self.submission.memory_limit, self.submission.meta
+        )
+
+        try:
+            self.grader = problem.grader_class(
+                self, problem, self.submission.language, utf8bytes(self.submission.source)
+            )
+        except CompileError as compilation_error:
+            error = compilation_error.args[0] or b'compiler exited abnormally'
+            yield IPC_COMPILE_ERROR, (error,)
+            return
+        else:
+            binary = self.grader.binary
+            if hasattr(binary, 'warning') and binary.warning is not None:
+                yield IPC_COMPILE_MESSAGE, (binary.warning,)
+
+        yield IPC_GRADING_BEGIN, (self.grader.is_pretested,)
+
+        flattened_cases = []
+        batch_number = 0
+        for case in self.grader.cases():
+            if isinstance(case, BatchedTestCase):
+                batch_number += 1
+                for batched_case in case.batched_cases:
+                    flattened_cases.append((batch_number, batched_case))
+            else:
+                flattened_cases.append((None, case))
+
+        case_number = 0
+        is_short_circuiting = False
+        for batch_number, cases in groupby(flattened_cases, key=itemgetter(0)):
+            if batch_number:
+                yield IPC_BATCH_BEGIN, (batch_number,)
+
+            for _, case in cases:
+                case_number += 1
+
+                # Stop grading if we're short circuiting
+                if is_short_circuiting:
+                    result = Result(case, result_flag=Result.SC)
+                else:
+                    result = self.grader.grade(case)
+
+                    # If the submission was killed due to an user-initiated abort, any result is meaningless.
+                    if self._abort_requested:
+                        yield IPC_GRADING_ABORTED, ()
+                        return
+
+                    # If the WA bit of result_flag is set and we are set to short-circuit (e.g., in a batch), short
+                    # circuit the rest of the cases. Do the same if the case is a pretest (i.e. has 0 points)
+                    if (result.result_flag & Result.WA) and (
+                        batch_number or not case.points or self.submission.short_circuit
+                    ):
+                        is_short_circuiting = True
+
+                yield IPC_RESULT, (batch_number, case_number, result)
+
+            if batch_number:
+                yield IPC_BATCH_END, (batch_number,)
+                is_short_circuiting &= self.submission.short_circuit
```

Comment:
I think `is_short_circuiting` shouldn't be unset if the submission failed on a 0-point case (Prod example: https://dmoj.ca/submission/2110231)

Reply:
(The current behaviour is different than the behaviour before this commit)

Reply:
Thanks, fixed, I believe. https://dmoj.ca/submission/2110247, https://github.com/DMOJ/judge-server/compare/43156ec1a8bad901f998df0df6edf949c4574135..2c0c2a2639bf56a20a261705d39243ed35bb63e2